pkg/controllers/tlspolicy/tlspolicy_controller.go

```diff
@@ -0,0 +1,271 @@
+	"github.com/Kuadrant/multicluster-gateway-controller/pkg/apis/v1alpha1"
+	"github.com/Kuadrant/multicluster-gateway-controller/pkg/controllers/events"
+)
+
+const (
+	TLSPolicyFinalizer           = "kuadrant.io/tls-policy"
+	TLSPoliciesBackRefAnnotation = "kuadrant.io/tlspolicies"
+	TLSPolicyBackRefAnnotation   = "kuadrant.io/tlspolicy"
+)
+
+type TLSPolicyRefsConfig struct{}
+
+func (c *TLSPolicyRefsConfig) PolicyRefsAnnotation() string {
+	return TLSPoliciesBackRefAnnotation
+}
+
+// TLSPolicyReconciler reconciles a TLSPolicy object
+type TLSPolicyReconciler struct {
+	reconcilers.TargetRefReconciler
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=kuadrant.io,resources=tlspolicies,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=kuadrant.io,resources=tlspolicies/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=kuadrant.io,resources=tlspolicies/finalizers,verbs=update
+
+func (r *TLSPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Logger().WithValues("TLSPolicy", req.NamespacedName)
+	log.Info("Reconciling TLSPolicy")
+	ctx = crlog.IntoContext(ctx, log)
+
+	previous := &v1alpha1.TLSPolicy{}
+	if err := r.Client().Get(ctx, req.NamespacedName, previous); err != nil {
+		if err := client.IgnoreNotFound(err); err == nil {
+			return ctrl.Result{}, nil
+		} else {
+			return ctrl.Result{}, err
+		}
+	}
+
+	tlsPolicy := previous.DeepCopy()
+	log.V(3).Info("TLSPolicyReconciler Reconcile", "tlsPolicy", tlsPolicy, "tlsPolicy.Spec", tlsPolicy.Spec)
+
+	markedForDeletion := tlsPolicy.GetDeletionTimestamp() != nil
+
+	targetNetworkObject, err := r.FetchValidTargetRef(ctx, tlsPolicy.GetTargetRef(), tlsPolicy.Namespace)
+	log.V(3).Info("TLSPolicyReconciler targetNetworkObject", "targetNetworkObject", targetNetworkObject)
+	if err != nil {
+		if !markedForDeletion {
+			if apierrors.IsNotFound(err) {
+				log.V(3).Info("Network object not found. Cleaning up")
+				err := r.deleteResources(ctx, tlsPolicy, nil)
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+			}
+			return ctrl.Result{}, err
+		}
+		targetNetworkObject = nil // we need the object set to nil when there's an error, otherwise deleting the resources (when marked for deletion) will panic
+	}
+
+	if markedForDeletion {
+		log.V(3).Info("cleaning up tls policy")
+		if controllerutil.ContainsFinalizer(tlsPolicy, TLSPolicyFinalizer) {
+			if err := r.deleteResources(ctx, tlsPolicy, targetNetworkObject); err != nil {
+				return ctrl.Result{}, err
+			}
+			if err := r.RemoveFinalizer(ctx, tlsPolicy, TLSPolicyFinalizer); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// add finalizer to the tlsPolicy
+	if !controllerutil.ContainsFinalizer(tlsPolicy, TLSPolicyFinalizer) {
+		if err := r.AddFinalizer(ctx, tlsPolicy, TLSPolicyFinalizer); client.IgnoreNotFound(err) != nil {
+			return ctrl.Result{Requeue: true}, err
```

Comment:
if you return an err as I understand it, it will auto get requeued? So do we need the Requeue true and the err?